lib/spack/spack/test/flag_mixing.py

```diff
@@ -0,0 +1,327 @@
+
+
+def test_two_dependents_flag_mixing(concretize_scope, test_repo):
+    root_spec1 = Spec("w~moveflaglater").concretized()
+    spec1 = root_spec1["y"]
+    assert spec1.compiler_flags["cflags"] == "-d0 -d1 -d2".split()
+
+    root_spec2 = Spec("w+moveflaglater").concretized()
+    spec2 = root_spec2["y"]
+    assert spec2.compiler_flags["cflags"] == "-d3 -d1 -d2".split()
+
+
+def test_propagate_and_compiler_cfg(concretize_scope, test_repo):
+    conf_str = _compiler_cfg_one_entry_with_cflags("-f2")
+    update_concretize_scope(conf_str, "compilers")
+
+    root_spec = Spec("v %gcc@12.100.100 cflags=='-f1'").concretized()
+    assert root_spec["y"].satisfies("cflags='-f1 -f2'")
+
+
+def test_propagate_and_spec_dep(concretize_scope, test_repo):
+    root_spec1 = Spec("v cflags=='-f1' ^y cflags='-f2'").concretized()
+    assert root_spec1["y"].satisfies("cflags='-f1 -f2'")
+
+
+def test_propagate_and_pkg_dep(concretize_scope, test_repo):
+    root_spec1 = Spec("x ~activatemultiflag cflags=='-f1'").concretized()
+    assert root_spec1["y"].satisfies("cflags='-f1 -d1'")
+
+
+def test_propagate_and_require(concretize_scope, test_repo):
+    conf_str = """\
+packages:
+  y:
+    require: cflags="-f2"
+"""
+    update_concretize_scope(conf_str, "packages")
+
+    root_spec1 = Spec("v cflags=='-f1'").concretized()
+    assert root_spec1["y"].satisfies("cflags='-f1 -f2'")
+
+    # Next, check that a requirement does not "undo" a request for
+    # propagation from the command-line spec
+    conf_str = """\
+packages:
+  v:
+    require: cflags="-f1"
+"""
+    update_concretize_scope(conf_str, "packages")
+
+    root_spec2 = Spec("v cflags=='-f1'").concretized()
+    assert root_spec2["y"].satisfies("cflags='-f1'")
+
+    # Note: requirements cannot enforce propagation: any attempt to do
+    # so will generate a concretization error; this likely relates to
+    # the note about #37180 in concretize.lp
+
+
+def test_dev_mix_flags(tmp_path, concretize_scope, mutable_mock_env_path, test_repo):
+    src_dir = tmp_path / "x-src"
+
+    env_content = f"""\
+spack:
+  specs:
+  - y %gcc@12.100.100 cflags=='-fsanitize=address'
+  develop:
+    y:
+      spec: y cflags=='-fsanitize=address'
+      path: {src_dir}
+"""
+
+    conf_str = _compiler_cfg_one_entry_with_cflags("-f1")
+    update_concretize_scope(conf_str, "compilers")
+
+    manifest_file = tmp_path / ev.manifest_name
+    manifest_file.write_text(env_content)
+    e = ev.create("test", manifest_file)
+    with e:
+        e.concretize()
+    e.write()
+
+    (result,) = list(j for i, j in e.concretized_specs() if j.name == "y")
+
+    assert result["y"].satisfies("cflags='-fsanitize=address -f1'")
+
+
+def test_diamond_dep_flag_mixing(concretize_scope, test_repo):
+    root_spec1 = Spec("t").concretized()
+    spec1 = root_spec1["y"]
+    assert spec1.satisfies('cflags="-c1 -c2 -d1 -d2 -e1 -e2"')
+    assert spec1.compiler_flags["cflags"] == "-c1 -c2 -e1 -e2 -d1 -d2".split()
```

Comment:
For this test (are there others here?) we need to write somewhere that the result being asserted depends on the implementation details of our topo-sorting.

Reply:
I added a docstring explaining this.